{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymongo\n",
    "from pymongo import MongoClient\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import all excel datasets using pandas\n",
    "# excel files for the first set of data\n",
    "# tsangkaris\n",
    "sum_table1 = pd.read_excel(\"Supplementary Table 8.xlsx\")\n",
    "CB_1 = pd.read_excel(\"Mouse_Brain_-_Cerebellum_7_weeks.xlsx\")\n",
    "CC_1 = pd.read_excel(\"Mouse_Brain_-_Cortex_7_weeks.xlsx\")\n",
    "HT_1 = pd.read_excel(\"mouse_brain_-_hipothalamus_7_weeks-01.xlsx\")\n",
    "HC_1 = pd.read_excel(\"mouse_brain_-_hipocampus_7_weeks-01.xlsx\")\n",
    "MD_1 = pd.read_excel(\"mouse_brain_-_medulla_7_weeks_old-01.xlsx\")\n",
    "MB_1 = pd.read_excel(\"Mouse_Brain_-_Mid_Brain_7_weeks.xlsx\")\n",
    "OB_1 = pd.read_excel(\"mouse_brain_-_olfactory_balb_7_weeks-01.xlsx\")\n",
    "\n",
    "# taraslia\n",
    "sum_table2 = pd.read_excel(\"TARASLIA et al TABLE 1.xls\", skiprows=1)\n",
    "CB_2 = pd.read_excel(\"TARASLIA et al TABLE 6.xls\", skiprows=1)\n",
    "CC_2 = pd.read_excel(\"TARASLIA et al TABLE 8.xls\", skiprows=1)\n",
    "HT_2 = pd.read_excel(\"TARASLIA et al TABLE 7.xls\", skiprows=1)\n",
    "HC_2 = pd.read_excel(\"TARASLIA et al TABLE 3.xls\", skiprows=1)\n",
    "MD_2 = pd.read_excel(\"TARASLIA et al TABLE 5.xls\", skiprows=1)\n",
    "MB_2 = pd.read_excel(\"TARASLIA et al TABLE 4.xls\", skiprows=1)\n",
    "OB_2 = pd.read_excel(\"TARASLIA et al TABLE 2.xls\", skiprows=1)\n",
    "\n",
    "MB_2.rename(columns={'Protein MW  (Da)':'Protein MW (Da)'}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First of all try to change the column names so that they are all the same for both origins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace in series 1 tables the accession with entry name\n",
    "mapping_df = sum_table1[['Entry', 'Entry name']].copy()\n",
    "\n",
    "dfs_series1 = [CB_1, CC_1, HT_1, HC_1, MD_1, MB_1, OB_1]\n",
    "\n",
    "for df in dfs_series1:\n",
    "    # Use the map function to replace values in the 'Accession' column\n",
    "    df['Accession'] = df['Accession'].map(mapping_df.set_index('Entry')['Entry name'])\n",
    "\n",
    "# rename the 'entry name' in the supplementary table 1\n",
    "sum_table1.rename(columns={'Entry name':'Accession name'}, inplace=True)\n",
    "sum_table1.drop('Entry', axis=1, inplace=True)\n",
    "\n",
    "sum_table2.rename(columns={'Accession Name':'Accession name'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create lists of the dataframes from the two different sources\n",
    "dfs_series1 = [CB_1, CC_1, HT_1, HC_1, MD_1, MB_1, OB_1]\n",
    "dfs_series2 = [CB_2, CC_2, HT_2, HC_2, MD_2, MB_2, OB_2]\n",
    "\n",
    "# convert Da to kDa in 2nd series of documents and rename the columns\n",
    "for df in dfs_series2:\n",
    "    df['Protein MW (Da)'] = df['Protein MW (Da)'] / 1000\n",
    "    df.rename(columns={'Accession Name':'Accession name',\n",
    "                       'Protein MW (Da)': 'Molecular weight (kDa)',\n",
    "                       'pI-value':'pI-Value'}, inplace=True)\n",
    "\n",
    "# renaming the column names of the first series\n",
    "for df in dfs_series1:\n",
    "    df.rename(columns={'Score': 'Mascot Score',\n",
    "                       'Accession': 'Accession name',\n",
    "                       'Coverage':'Coverage (%)',\n",
    "                       '# Proteins': 'Proteins',\n",
    "                       '# Unique Peptides': 'Unique Peptides',\n",
    "                       '# Peptides':'Peptides',\n",
    "                       '# PSMs':'PSMs',\n",
    "                       '# AAs': 'Amino Acids',\n",
    "                       'MW [kDa]': 'Molecular weight (kDa)',\n",
    "                       'calc. pI': 'pI-Value'}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Cerebral cortex</th>\n",
       "      <th>Olfactory Bulb</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>Hypothalamus</th>\n",
       "      <th>Mid brain</th>\n",
       "      <th>Cerebellum</th>\n",
       "      <th>Medulla</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A067XG52_MOUSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUTS2_MOUSE</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A4W9_MOUSE</td>\n",
       "      <td>Neuronal growth regulator 1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SMKY_MOUSE</td>\n",
       "      <td>Sperm motility kinase Y (EC 2.7.11.1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0JLR7_MOUSE</td>\n",
       "      <td>Ahnak protein (Fragment)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Accession name                                 Protein names  \\\n",
       "0  A0A067XG52_MOUSE                                           NaN   \n",
       "1       AUTS2_MOUSE  Autism susceptibility gene 2 protein homolog   \n",
       "2      A0A4W9_MOUSE                   Neuronal growth regulator 1   \n",
       "3        SMKY_MOUSE         Sperm motility kinase Y (EC 2.7.11.1)   \n",
       "4      A0JLR7_MOUSE                      Ahnak protein (Fragment)   \n",
       "\n",
       "   Cerebral cortex  Olfactory Bulb  Hippocampus  Hypothalamus  Mid brain  \\\n",
       "0              0.0             0.0          0.0           0.0        1.0   \n",
       "1              0.0             0.0          1.0           0.0        0.0   \n",
       "2              0.0             0.0          0.0           0.0        1.0   \n",
       "3              0.0             0.0          0.0           1.0        0.0   \n",
       "4              0.0             0.0          0.0           0.0        1.0   \n",
       "\n",
       "   Cerebellum  Medulla  \n",
       "0         0.0      0.0  \n",
       "1         0.0      0.0  \n",
       "2         0.0      0.0  \n",
       "3         0.0      0.0  \n",
       "4         0.0      0.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_table1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession name</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>Protein MW</th>\n",
       "      <th>pI-Value</th>\n",
       "      <th>Brain part</th>\n",
       "      <th>Subcellular localisation</th>\n",
       "      <th>Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1433B_MOUSE</td>\n",
       "      <td>14-3-3 protein beta/alpha</td>\n",
       "      <td>28183</td>\n",
       "      <td>4.62</td>\n",
       "      <td>OB, HT, MD</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>Transcription rergulation, cytoplasmic sequest...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1433E_MOUSE</td>\n",
       "      <td>14-3-3 protein epsilon</td>\n",
       "      <td>29326</td>\n",
       "      <td>4.50</td>\n",
       "      <td>OB, MB, HT, HC, CB, MD</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>Cerebral cortex development, hippocampus devel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1433F_MOUSE</td>\n",
       "      <td>14-3-3 protein eta</td>\n",
       "      <td>28365</td>\n",
       "      <td>4.66</td>\n",
       "      <td>OB</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>Glucocorticoid catabolism, glucocorticoid rece...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1433G_MOUSE</td>\n",
       "      <td>14-3-3 protein gamma</td>\n",
       "      <td>28456</td>\n",
       "      <td>4.70</td>\n",
       "      <td>OB, CC, MB, HT, HC, CB, MD</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>Neuronal differentiation, regulation of synapt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1433T_MOUSE</td>\n",
       "      <td>14-3-3 protein theta</td>\n",
       "      <td>28046</td>\n",
       "      <td>4.54</td>\n",
       "      <td>OB, MD</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>Trasnscription, signal transduction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Accession name                Protein name  Protein MW  pI-Value  \\\n",
       "0    1433B_MOUSE  14-3-3 protein beta/alpha        28183      4.62   \n",
       "1    1433E_MOUSE     14-3-3 protein epsilon        29326      4.50   \n",
       "2    1433F_MOUSE         14-3-3 protein eta        28365      4.66   \n",
       "3    1433G_MOUSE       14-3-3 protein gamma        28456      4.70   \n",
       "4    1433T_MOUSE       14-3-3 protein theta        28046      4.54   \n",
       "\n",
       "                   Brain part Subcellular localisation  \\\n",
       "0                  OB, HT, MD                Cytoplasm   \n",
       "1      OB, MB, HT, HC, CB, MD                Cytoplasm   \n",
       "2                          OB                Cytoplasm   \n",
       "3  OB, CC, MB, HT, HC, CB, MD                Cytoplasm   \n",
       "4                      OB, MD                Cytoplasm   \n",
       "\n",
       "                                            Function  \n",
       "0  Transcription rergulation, cytoplasmic sequest...  \n",
       "1  Cerebral cortex development, hippocampus devel...  \n",
       "2  Glucocorticoid catabolism, glucocorticoid rece...  \n",
       "3  Neuronal differentiation, regulation of synapt...  \n",
       "4                Trasnscription, signal transduction  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_table2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if all accession names of each brain region exist in the supplementary table\n",
    "\n",
    "# create a list out of the accession names of the supplementary tables\n",
    "prots1 = sum_table1['Accession name'].tolist()\n",
    "\n",
    "for df in dfs_series1:\n",
    "    # create a list out of the accession names of each df\n",
    "    accession_list = df['Accession name'].tolist()\n",
    "\n",
    "    # check that all proteins from each brain region exists in the supplementary table\n",
    "    # if not return the index, and the data frame name\n",
    "    for i, accession_name in enumerate(accession_list):\n",
    "        if accession_name not in prots1:\n",
    "            print(f\"Accession name '{accession_name}' at index {i} in DataFrame {df.name} does not exist in the supplementary table.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accession name 'CV028_MOUSE' at index 5 in DataFrame 1 does not exist in the supplementary table.\n",
      "Accession name 'CX4NB_MOUSE' at index 2 in DataFrame 5 does not exist in the supplementary table.\n",
      "Accession name 'CA093_MOUSE' at index 5 in DataFrame 6 does not exist in the supplementary table.\n"
     ]
    }
   ],
   "source": [
    "prots2 = sum_table2['Accession name'].tolist()\n",
    "\n",
    "for df_index, df in enumerate(dfs_series2):\n",
    "    # Check if 'Accession name' column exists in the DataFrame\n",
    "    if 'Accession name' in df.columns:\n",
    "        accession_list = df['Accession name'].tolist()\n",
    "\n",
    "        # Check that all proteins from each brain region exist in the supplementary table\n",
    "        # If not, return the index, and the data frame index\n",
    "        for i, accession_name in enumerate(accession_list):\n",
    "            if accession_name not in prots2:\n",
    "                print(f\"Accession name '{accession_name}' at index {i} in DataFrame {df_index} does not exist in the supplementary table.\")\n",
    "    else:\n",
    "        print(f\"DataFrame {df_index} does not have the 'Accession name' column.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/4b/f4zzvdd97mdb6v7rmbgvl76w0000gn/T/ipykernel_8904/3656710357.py:3: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  sum_table2 = sum_table2.append({'Accession name':'CV028_MOUSE', 'Brain part': 'CC'}, ignore_index=True)\n",
      "/var/folders/4b/f4zzvdd97mdb6v7rmbgvl76w0000gn/T/ipykernel_8904/3656710357.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  sum_table2 = sum_table2.append({'Accession name':'CX4NB_MOUSE', 'Brain part': 'MB'}, ignore_index=True)\n",
      "/var/folders/4b/f4zzvdd97mdb6v7rmbgvl76w0000gn/T/ipykernel_8904/3656710357.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  sum_table2 = sum_table2.append({'Accession name':'CA093_MOUSE', 'Brain part': 'OB'}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# Add those accession names in the supplementary table\n",
    "\n",
    "sum_table2 = sum_table2.append({'Accession name':'CV028_MOUSE', 'Brain part': 'CC'}, ignore_index=True)\n",
    "sum_table2 = sum_table2.append({'Accession name':'CX4NB_MOUSE', 'Brain part': 'MB'}, ignore_index=True)\n",
    "sum_table2 = sum_table2.append({'Accession name':'CA093_MOUSE', 'Brain part': 'OB'}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Three accession names existed in the 6 regions but didn't exist in the Summarized table in the Taraslia data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein and Brain regions Collections\n",
    "Try out the generation of unique protein accession_name collection AND keep in dataframe which are unique from which origin, and which are common as well. Also, create a collection for the unique brain regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession name</th>\n",
       "      <th>Origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A067XG52_MOUSE</td>\n",
       "      <td>Tsagkaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUTS2_MOUSE</td>\n",
       "      <td>Tsagkaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A4W9_MOUSE</td>\n",
       "      <td>Tsagkaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SMKY_MOUSE</td>\n",
       "      <td>Tsagkaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0JLR7_MOUSE</td>\n",
       "      <td>Tsagkaris</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Accession name     Origin\n",
       "0  A0A067XG52_MOUSE  Tsagkaris\n",
       "1       AUTS2_MOUSE  Tsagkaris\n",
       "2      A0A4W9_MOUSE  Tsagkaris\n",
       "3        SMKY_MOUSE  Tsagkaris\n",
       "4      A0JLR7_MOUSE  Tsagkaris"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a new dataframe that will contain which proteins are unique to which origin (series 1: tsangkaris, series 2: taraslia)\n",
    "proteins = prots1 + prots2\n",
    "\n",
    "proteins_all = pd.DataFrame({'Accession name':pd.Series(proteins).unique(), 'Origin': None})\n",
    "prots1_series = pd.Series(prots1)\n",
    "prots2_series = pd.Series(prots2)\n",
    "\n",
    "# identify common, and unique proteins with respect to each origin\n",
    "common_prots = prots1_series[prots1_series.isin(prots2_series)]\n",
    "unique_prots1 = prots1_series[~prots1_series.isin(common_prots)]\n",
    "unique_prots2 = prots2_series[~prots2_series.isin(common_prots)]\n",
    "\n",
    "# loop through each row in the DataFrame\n",
    "for index, row in proteins_all.iterrows():\n",
    "    accession_name = row['Accession name']\n",
    "    \n",
    "    # Check the group to which the 'Accession name' belongs and update the 'Origin' column\n",
    "    if accession_name in common_prots.values:\n",
    "        proteins_all.at[index, 'Origin'] = ['Tsagkaris', 'Taraslia']\n",
    "    elif accession_name in unique_prots1.values:\n",
    "        proteins_all.at[index, 'Origin'] = 'Tsagkaris'\n",
    "    elif accession_name in unique_prots2.values:\n",
    "        proteins_all.at[index, 'Origin'] = 'Taraslia'\n",
    "\n",
    "\n",
    "proteins_all.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
